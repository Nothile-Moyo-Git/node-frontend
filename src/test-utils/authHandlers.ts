// handlers.ts
import { graphql } from "msw";

export const handlers = [
  graphql.query("PostUserDetailsResponse", (req, res, ctx) => {
    return res(
      ctx.data({
        PostUserDetailsResponse: {
          sessionCreated: "2024-01-01",
          sessionExpires: "2024-12-31",
          success: true,
          user: {
            _id: "1",
            name: "Nothile Moyo",
            email: "nothile@example.com",
            password: "secret",
            confirmPassword: "secret",
            status: "active",
            posts: [],
          },
        },
      }),
    );
  }),
];
