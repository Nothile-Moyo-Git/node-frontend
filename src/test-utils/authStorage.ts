/**
 * Date Created: 05/08/2025
 *
 * Author: Nothile Moyo
 *
 * Description: This is the local storage handler file for my tests
 * It will handle anything that needs to be set in local storage, in this case, authentication
 *
 * All local storage items are deleted when the request is complete
 */

/**
 * @name setMockAuthStorage
 *
 * @description Mocks setting authentication data for a user
 *
 * @type () => void
 */
export const setMockAuthStorage = ({
  token = "mock-token",
  userId = "mock-user-id",
  expiresIn = new Date(Date.now() + 3600000).toISOString(),
} = {}) => {
  localStorage.setItem("token", token);
  localStorage.setItem("userId", userId);
  localStorage.setItem("expiresIn", expiresIn);
};

export const clearAuthStorage = () => {
  localStorage.removeItem("token");
  localStorage.removeItem("userId");
  localStorage.removeItem("expiresIn");
};
