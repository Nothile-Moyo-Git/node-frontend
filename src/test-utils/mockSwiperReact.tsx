/**
 * Date created : 01/10/2025
 * Author : Nothile Moyo
 *
 * Test routing file, this will create routes which can be used in Jest for testing
 * We use this to be able to render our components
 *
 */

import { ReactElement } from "react";

export const Swiper = (children: ReactElement) => (
  <div data-testid="swiper">{children}</div>
);

export const SwiperSlide = (children: ReactElement) => {
  <div data-testid="swiper-slide">{children}</div>;
};

export default {
  Swiper,
  SwiperSlide,
};
