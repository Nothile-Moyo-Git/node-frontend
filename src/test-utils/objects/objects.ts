import { User } from "../../@types";
import { ContextProps } from "../../context/AppContext";
import { Post } from "../../@types";

// Define a user here which should have their details rendered on the main App page
export const mockUser: User = {
  _id: "6624158268f8cf47bea31396",
  name: "Nothile Moyo",
  email: "nothile1@gmail.com",
  password: "test",
  status: "active",
  posts: [
    "662423764e8c8b1633534be8",
    "662423884e8c8b1633534bf0",
    "662e7bcdd94fde36bf4bb554",
    "662e7c6ad94fde36bf4bb55c",
    "67843561d02db477bac4843b",
  ],
};

export const mockContext: ContextProps = {
  baseUrl: process.env.API_URL_DEV ?? "http://localhost:4000",
  checkAuthentication: () => true,
  logoutUser: () => {},
  token: "fake-token",
  userId: "12345-nothile-id",
  userAuthenticated: true,
  validateAuthentication: () => {},
};

// Mocking our posts
export const mockPosts: Post[] = [
  {
    _id: "662423764e8c8b1633534be8",
    fileLastUpdated: "",
    fileName: "2B.png",
    title: "2B",
    imageUrl: "..\frontend\x02src\x02images\x022B.png",
    content:
      "YoRHa No. 2 Type B,[a] commonly known as 2B,[b] is a fictional android from the 2017 video game Nier: Automata, a spin-off of the Drakengard series developed by PlatinumGames and published by Square Enix. One of the game's three protagonists, 2B is a soldier for YoRHa, an android task force fighting a proxy war with alien-created Machine Lifeforms. Her story arc focuses on her backstory within YoRHa, and her relationship with her partner 9S, a reconnaissance android. She is also featured in related media, such as the anime Nier: Automata Ver1.1a.",
    creator: "12345-nothile-id",
    createdAt: "1713644407010",
    updatedAt: "1713644407010",
  },
  {
    _id: "662423884e8c8b1633534bf0",
    fileLastUpdated: "",
    fileName: "Alfira-face.jpg",
    title: "Alfira",
    imageUrl: "..\x02frontend\x02src\x02images\x02Alfira-face.jpg",
    content:
      "Alfira is part of the group of tiefling refugees who were exiled from Elturel, and Bardic apprentice to a woman named Lihala. On the journey to the Emerald Grove, they were attacked by gnolls, and Lihala was killed. Alfira has since been trying to compose a eulogy for her.",
    creator: "12345-nothile-id",
    createdAt: "1713644424070",
    updatedAt: "1713644424070",
  },
  {
    _id: "662e7bcdd94fde36bf4bb554",
    fileLastUpdated: "",
    fileName: "Shanalotte.jpg",
    title: "Emerald Herald",
    imageUrl: "..\x02frontend\x02src\x02images\x02Shanalotte.jpg",
    content:
      "The Emerald Herald is first encountered in the town of Majula, where she sits beside the cliff overlooking the ocean. Her true nature and purpose are shrouded in mystery, but she plays a crucial role in the player's quest by providing them with the means to level up and upgrade their Estus Flask.",
    creator: "12345-nothile-id",
    createdAt: "1714322381968",
    updatedAt: "1714322381968",
  },
  {
    _id: "662e7c6ad94fde36bf4bb55c",
    fileLastUpdated: "",
    fileName: "Tiefling.jpg",
    title: "Tieflings",
    imageUrl: "..\x02frontend\x02src\x02images\x02Tiefling.jpg",
    content:
      "But you do see the way people look at you, devilâ€™s child? Those black eyes, cold as a winter storm, were staring right into her heart and the sudden seriousness in his voice jolted her.",
    creator: "12345-nothile-id",
    createdAt: "1714322538849",
    updatedAt: "1741993615409",
  },
  {
    _id: "67843561d02db477bac4843b",
    fileLastUpdated: "",
    fileName: "Edelgard.jpg",
    title: "The Sixth Post",
    imageUrl: "../frontend/src/images/Edelgard.jpg",
    content: "Descriptions are a useful thing when you have them!",
    creator: "12345-nothile-id",
    createdAt: "1736717665449",
    updatedAt: "1756408330943",
  },
  {
    _id: "68b0aac75d6f6df6ef698e7e",
    fileLastUpdated: "",
    fileName: "Shanalotte.jpg",
    title: "Testing post number 6",
    imageUrl: "../frontend/src/images/Shanalotte.jpg",
    content: "Testing post number 6 for functionality",
    creator: "12345-nothile-id",
    createdAt: "1756408519264",
    updatedAt: "1756495750074",
  },
];
