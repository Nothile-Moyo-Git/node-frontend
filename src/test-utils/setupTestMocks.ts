/**
 * Date created: 30/07/2025
 *
 * Author: Nothile
 *
 * Methods test file, this file defines reusable methods and hooks which can be used across my tests
 * These methods are intended to be used to replicate the official functionality of the app
 *
 */

import { useState } from "react";
import { User } from "../@types";

// Mock useNavigate so it works with jest
export const mockUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockUsedNavigate,
}));

// This allows us to be able to mock hooks in Jest otherwise we use the actual hook which triggers errors
jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

// Cast since we're using a ts file and we want to manipulate this
const mockedUseState = useState as jest.Mock;

const setMockState = jest.fn();

// Default mock implementation of usestate
mockedUseState.mockImplementation((init) => [init, setMockState]);

// Utility function to update mocked isLoading in tests
export const setAppStateMock = (
  isLoading: boolean,
  loadingError: boolean,
  user: User | undefined,
  sessionExpiryDate: string | undefined,
  sessionCreationDate: string | undefined,
) => {
  // Call the state in the same order but ensure that our initial values aren't set multiple times
  let hasSetLoading = false;
  let hasSetUser = false;
  const hasSetExpiry = false;

  mockedUseState.mockImplementation((init) => {
    if (typeof init === "boolean" && !hasSetLoading) {
      hasSetLoading = true;
      return [isLoading, setMockState];
    }
    if (typeof init === "boolean") {
      return [loadingError, setMockState];
    }
    if (init === undefined && !hasSetUser) {
      hasSetUser = true;
      return [user, setMockState];
    }
    /* if (typeof init === "string" && !hasSetExpiry) {
      hasSetExpiry = true;
      return [sessionExpiryDate, setMockState];
    }
    if (typeof init === "string") {
      return [sessionCreationDate, setMockState];
    } */
    return [init, setMockState];
  });
};
