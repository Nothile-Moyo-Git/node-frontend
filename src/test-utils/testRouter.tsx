/**
 * Date created : 30/07/2025
 * Author : Nothile Moyo
 *
 * Test routing file, this will create routes which can be used in Jest for testing
 * We use this to be able to render our components
 *
 */
import { ReactNode } from "react";
import { render } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";

export function renderWithRouter(ui: ReactNode, { route = "/" } = {}, context) {
  const router = createMemoryRouter(
    [
      {
        path: "/",
        element: ui,
      },
    ],
    { initialEntries: [route] },
  );

  return render(<RouterProvider router={router} />);
}
