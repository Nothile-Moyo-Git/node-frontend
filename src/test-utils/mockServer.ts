import { setupServer } from "msw/node";
import { handlers } from "./authHandlers";

export const server = setupServer(...handlers);
