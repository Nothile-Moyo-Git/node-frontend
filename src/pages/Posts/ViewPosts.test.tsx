/**
 * Author: Nothile Moyo
 * Date created: 29/08/2025
 *
 * @description: This is the test file for the View Posts component
 * We authenticate the user, and if valid, we then show a list of all posts on the page
 */

import "@testing-library/jest-dom";
import {
  clearAuthStorage,
  setMockAuthStorage,
} from "../../test-utils/authStorage";
import { server } from "../../test-utils/mockServer";
import { renderWithAct, renderWithContext } from "../../test-utils/testRouter";
import { act, screen } from "@testing-library/react";
import { ViewPosts } from "./ViewPosts";
import { mockContext, mockPosts } from "../../test-utils/objects/objects";
import userEvent from "@testing-library/user-event";

// Mocking socket.io jest so we don't make a real connection
jest.mock("socket.io-client", () => {
  return {
    io: () => ({
      on: jest.fn(),
      emit: jest.fn(),
      removeAllListeners: jest.fn(),
    }),
  };
});

// Setup mocks and environment
beforeAll(() => {
  server.listen();
});

beforeEach(() => {
  setMockAuthStorage();
});

// Cleanup mocks and environment
afterEach(() => {
  server.resetHandlers();
  clearAuthStorage();
});

afterAll(() => {
  server.close();
});

describe("View Posts component", () => {
  it("Should match snapshot", async () => {
    global.fetch = jest.fn().mockResolvedValue({
      status: 200,
      ok: true,
      json: () =>
        Promise.resolve({
          data: {
            GetPostsResponse: {
              success: true,
              numberOfPages: 2,
              posts: mockPosts,
              message: "200 : Request was successful",
            },
          },
        }),
    });

    // Render the component with our posts and the images being mocked
    await act(async () => {
      renderWithContext(<ViewPosts />, { route: "/posts" }, mockContext);
    });

    // Check if the view posts component is rendered and we navigate to it successfully
    const appComponent = await screen.findByTestId("test-id-view-posts");
    expect(appComponent).toBeInTheDocument();
    expect(appComponent).toMatchSnapshot();
  });

  it("Should show loading spinner on initial render", async () => {
    // Make the request never resolve so the loading spinner keeps spinning
    global.fetch = jest.fn().mockImplementation(
      () =>
        new Promise(() => {
          // Never resolves â†’ keeps isLoading=true
        }),
    ) as jest.Mock;

    renderWithContext(<ViewPosts />, { route: "/posts" }, mockContext);

    const loadingIndicator = await screen.findByTestId(
      "test-id-loading-spinner",
    );
    expect(loadingIndicator).toBeVisible();
  });

  it("Renders the View Posts component successfully", async () => {
    global.fetch = jest.fn().mockResolvedValue({
      status: 200,
      ok: true,
      json: () =>
        Promise.resolve({
          data: {
            GetPostsResponse: {
              success: true,
              numberOfPages: 2,
              posts: mockPosts,
              message: "200 : Request was successful",
            },
          },
        }),
    });

    await act(async () => {
      renderWithContext(<ViewPosts />, { route: "/posts" }, mockContext);
    });

    // Check if the view posts component is rendered and we navigate to it successfully
    const appComponent = screen.getByTestId("test-id-view-posts");
    expect(appComponent).toBeInTheDocument();

    // Check to see if our posts exist
    // We're checking with different identifiers
    const post2B = screen.getByTestId(`test-id-post-${mockPosts[0]._id}`);
    expect(post2B).toBeVisible();

    const postAlfira = screen.getByText(mockPosts[1].title);
    expect(postAlfira).toBeVisible();

    const emeraldHerald = screen.getByText(mockPosts[2].title);
    expect(emeraldHerald).toBeVisible();
  });

  it("Renders the page with pagination and goes to the next page", async () => {
    global.fetch = jest.fn().mockResolvedValue({
      status: 200,
      ok: true,
      json: () =>
        Promise.resolve({
          data: {
            GetPostsResponse: {
              success: true,
              numberOfPages: 2,
              posts: mockPosts,
              message: "200 : Request was successful",
            },
          },
        }),
    });

    await renderWithAct(<ViewPosts />, { route: "/posts" }, mockContext);

    // Go to page 2 as we have 6 posts which allows us to use pagination
    const paginationPage2 = await screen.findByTestId(
      "test-id-pagination-page-2",
    );

    await act(async () => {
      userEvent.click(paginationPage2);
    });

    const post2B = await screen.findByTestId(
      `test-id-post-${mockPosts[3]._id}`,
    );
    expect(post2B).toBeVisible();

    const postAlfira = await screen.findByText(mockPosts[4].title);
    expect(postAlfira).toBeVisible();

    const emeraldHerald = await screen.findByText(mockPosts[5].title);
    expect(emeraldHerald).toBeVisible();
  });

  it("Deletes a post on the posts page", async () => {
    global.fetch = jest.fn().mockResolvedValue({
      status: 200,
      ok: true,
      json: () =>
        Promise.resolve({
          data: {
            GetPostsResponse: {
              success: true,
              numberOfPages: 2,
              posts: mockPosts,
              message: "200 : Request was successful",
            },
          },
        }),
    });

    await renderWithAct(<ViewPosts />, { route: "/posts" }, mockContext);
  });
});
