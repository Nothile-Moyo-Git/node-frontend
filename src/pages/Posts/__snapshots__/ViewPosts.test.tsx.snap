// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Posts component Should match snapshot 1`] = `
<section
  class="viewPosts"
  data-testid="test-id-view-posts"
>
  <h1
    class="viewPosts__title"
  >
    Posts
  </h1>
  <ul
    class="viewPosts__posts-list"
  >
    <li>
      <article
        class="article"
        data-testid="test-id-post-662423764e8c8b1633534be8"
      >
        <img
          alt="2B"
          class="article__image"
          draggable="false"
          src="test-file-stub"
        />
        <div
          class="article__content"
        >
          <h2
            class="article__title"
          >
            2B
          </h2>
          <p
            class="article__description"
          >
            YoRHa No. 2 Type B,[a] commonly known as 2B,[b] is a fictional android from the 2017 video game Nier: Automata, a spin-off of the Drakengard series developed by PlatinumGames and published by Square Enix. One of the game's three protagonists, 2B is a soldier for YoRHa, an android task force fighting a proxy war with alien-created Machine Lifeforms. Her story arc focuses on her backstory within YoRHa, and her relationship with her partner 9S, a reconnaissance android. She is also featured in related media, such as the anime Nier: Automata Ver1.1a.
          </p>
          <p>
            Uploaded: 2024/04/20 21:20:07
          </p>
          <div
            class="article__buttons"
          >
            <a
              class="link__read-more"
              href="/post/662423764e8c8b1633534be8"
            >
              Read more
            </a>
          </div>
        </div>
      </article>
    </li>
    <li>
      <article
        class="article"
        data-testid="test-id-post-662423884e8c8b1633534bf0"
      >
        <img
          alt="Alfira"
          class="article__image"
          draggable="false"
          src="test-file-stub"
        />
        <div
          class="article__content"
        >
          <h2
            class="article__title"
          >
            Alfira
          </h2>
          <p
            class="article__description"
          >
            Alfira is part of the group of tiefling refugees who were exiled from Elturel, and Bardic apprentice to a woman named Lihala. On the journey to the Emerald Grove, they were attacked by gnolls, and Lihala was killed. Alfira has since been trying to compose a eulogy for her.
          </p>
          <p>
            Uploaded: 2024/04/20 21:20:24
          </p>
          <div
            class="article__buttons"
          >
            <a
              class="link__read-more"
              href="/post/662423884e8c8b1633534bf0"
            >
              Read more
            </a>
          </div>
        </div>
      </article>
    </li>
    <li>
      <article
        class="article"
        data-testid="test-id-post-662e7bcdd94fde36bf4bb554"
      >
        <img
          alt="Emerald Herald"
          class="article__image"
          draggable="false"
          src="test-file-stub"
        />
        <div
          class="article__content"
        >
          <h2
            class="article__title"
          >
            Emerald Herald
          </h2>
          <p
            class="article__description"
          >
            The Emerald Herald is first encountered in the town of Majula, where she sits beside the cliff overlooking the ocean. Her true nature and purpose are shrouded in mystery, but she plays a crucial role in the player's quest by providing them with the means to level up and upgrade their Estus Flask.
          </p>
          <p>
            Uploaded: 2024/04/28 17:39:41
          </p>
          <div
            class="article__buttons"
          >
            <a
              class="link__read-more"
              href="/post/662e7bcdd94fde36bf4bb554"
            >
              Read more
            </a>
          </div>
        </div>
      </article>
    </li>
    <li>
      <article
        class="article"
        data-testid="test-id-post-662e7c6ad94fde36bf4bb55c"
      >
        <img
          alt="Tieflings"
          class="article__image"
          draggable="false"
          src="test-file-stub"
        />
        <div
          class="article__content"
        >
          <h2
            class="article__title"
          >
            Tieflings
          </h2>
          <p
            class="article__description"
          >
            But you do see the way people look at you, devilâ€™s child? Those black eyes, cold as a winter storm, were staring right into her heart and the sudden seriousness in his voice jolted her.
          </p>
          <p>
            Uploaded: 2024/04/28 17:42:18
          </p>
          <div
            class="article__buttons"
          >
            <a
              class="link__read-more"
              href="/post/662e7c6ad94fde36bf4bb55c"
            >
              Read more
            </a>
          </div>
        </div>
      </article>
    </li>
    <li>
      <article
        class="article"
        data-testid="test-id-post-67843561d02db477bac4843b"
      >
        <img
          alt="The Sixth Post"
          class="article__image"
          draggable="false"
          src="test-file-stub"
        />
        <div
          class="article__content"
        >
          <h2
            class="article__title"
          >
            The Sixth Post
          </h2>
          <p
            class="article__description"
          >
            Descriptions are a useful thing when you have them!
          </p>
          <p>
            Uploaded: 2025/01/12 21:34:25
          </p>
          <div
            class="article__buttons"
          >
            <a
              class="link__read-more"
              href="/post/67843561d02db477bac4843b"
            >
              Read more
            </a>
          </div>
        </div>
      </article>
    </li>
    <li>
      <article
        class="article"
        data-testid="test-id-post-68b0aac75d6f6df6ef698e7e"
      >
        <img
          alt="Testing post number 6"
          class="article__image"
          draggable="false"
          src="test-file-stub"
        />
        <div
          class="article__content"
        >
          <h2
            class="article__title"
          >
            Testing post number 6
          </h2>
          <p
            class="article__description"
          >
            Testing post number 6 for functionality
          </p>
          <p>
            Uploaded: 2025/08/28 20:15:19
          </p>
          <div
            class="article__buttons"
          >
            <a
              class="link__read-more"
              href="/post/68b0aac75d6f6df6ef698e7e"
            >
              Read more
            </a>
          </div>
        </div>
      </article>
    </li>
  </ul>
  <div
    class="paginator"
  >
    <button
      class="paginator__button paginator__button--current-normal"
      disabled=""
    >
      1
    </button>
    <a
      class="paginator__button"
      data-testid="test-id-pagination-page-2"
      href="/posts/2"
    >
      2
    </a>
  </div>
</section>
`;
